kernel/src/net/socket/unix/stream/connected.rs

```diff
@@ -1,65 +1,74 @@
 // SPDX-License-Identifier: MPL-2.0
 
-use core::ops::Deref;
-
-use ostd::sync::PreemptDisabled;
-
 use crate::{
     events::IoEvents,
-    fs::utils::{Channel, Consumer, Producer},
+    fs::utils::{ChannelPollee, Peered},
     net::socket::{
         unix::{addr::UnixSocketAddrBound, UnixSocketAddr},
         util::SockShutdownCmd,
     },
     prelude::*,
-    process::signal::{PollHandle, Pollee},
-    util::{MultiRead, MultiWrite},
+    process::signal::Pollee,
+    util::{
+        ring_buffer::{RbConsumer, RbProducer, RingBuffer},
+        MultiRead, MultiWrite,
+    },
 };
 
 pub(super) struct Connected {
-    addr: AddrView,
-    reader: Consumer<u8>,
-    writer: Producer<u8>,
+    inner: Peered<Inner>,
+    reader: Mutex<RbConsumer<u8>>,
+    writer: Mutex<RbProducer<u8>>,
 }
 
 impl Connected {
     pub(super) fn new_pair(
         addr: Option<UnixSocketAddrBound>,
         peer_addr: Option<UnixSocketAddrBound>,
-        reader_pollee: Option<Pollee>,
-        writer_pollee: Option<Pollee>,
+        is_read_shutdown: bool,
+        is_write_shutdown: bool,
+        pollee: Pollee,
     ) -> (Connected, Connected) {
-        let (writer_peer, reader_this) =
-            Channel::with_capacity_and_pollees(DEFAULT_BUF_SIZE, None, reader_pollee).split();
-        let (writer_this, reader_peer) =
-            Channel::with_capacity_and_pollees(DEFAULT_BUF_SIZE, writer_pollee, None).split();
+        pollee.invalidate();
+
+        let (this_writer, peer_reader) = RingBuffer::new(DEFAULT_BUF_SIZE).split();
+        let (peer_writer, this_reader) = RingBuffer::new(DEFAULT_BUF_SIZE).split();
+
+        let this_inner = Inner {
+            addr: SpinLock::new(addr),
+            pollee: ChannelPollee::with_pollee(pollee, is_write_shutdown),
+        };
+        let peer_inner = Inner {
+            addr: SpinLock::new(peer_addr),
+            pollee: ChannelPollee::with_pollee(Pollee::new(), is_read_shutdown),
+        };
 
-        let (addr_this, addr_peer) = AddrView::new_pair(addr, peer_addr);
+        let (this_inner, peer_inner) = Peered::new_pair(this_inner, peer_inner);
 
         let this = Connected {
-            addr: addr_this,
-            reader: reader_this,
-            writer: writer_this,
+            inner: this_inner,
+            reader: Mutex::new(this_reader),
+            writer: Mutex::new(this_writer),
```

Comment:
Is it possible to directly store reader and writer in `Inner`? Then we may can just let Connected a type alias of `Peered`?

```rust
let this_inner = Inner {
   addr: SpinLock::new(addr),
   pollee: ChannelPollee::with_pollee(pollee, is_write_shutdown),
   reader: Mutex::new(this_reader),
   writer: Mutex::new(this_writer),
};
```

But the `Drop` of `Connected` seems to be a problem....

Reply:
Well, we may do this in the future if we want to resize the ring buffer, so we'll have to access the reader/writer in the remote endpoint.

Currently, it seems that an endpoint should never access the remote reader/writer, so I think having it in `Connected` is the right location?

Reply:
Sounds reasonable. You can keep the current location. It's acceptable to me.